p2p/module.go

```diff
@@ -96,13 +120,31 @@ func (m *p2pModule) Start() error {
 			telemetry.P2P_NODE_STARTED_TIMESERIES_METRIC_DESCRIPTION,
 		)
 
-	addrBook, err := ValidatorMapToAddrBook(m.p2pCfg, m.bus.GetConsensusModule().ValidatorMap())
+	currentHeight := m.bus.GetConsensusModule().CurrentHeight()
+	var (
+		addrBook typesP2P.AddrBook
+		err      error
+	)
+
+	if m.GetBus().GetPersistenceModule() == nil {
+		// we are getting called by the client and we use the "legacy behaviour"
+		// TODO (team): improve this.
+		addrBook, err = ValidatorMapToAddrBook(m.p2pCfg, m.bus.GetConsensusModule().ValidatorMap())
+	} else {
+		addrBook, err = m.getAddrBookPerHeight(currentHeight)
+	}
 	if err != nil {
 		return err
 	}
 
 	if m.p2pCfg.GetUseRainTree() {
-		m.network = raintree.NewRainTreeNetwork(m.address, addrBook)
+		if m.GetBus().GetPersistenceModule() == nil {
+			// we are getting called by the client and we use the "legacy behaviour"
+			// TODO (team): improve this.
+			m.network = raintree.NewRainTreeNetwork(m.address, addrBook, m.p2pCfg)
+		} else {
+			m.network = raintree.NewRainTreeNetworkWithAddrBookProvider(m.address, m.getAddrBookPerHeight, currentHeight, m.p2pCfg)
```

Comment:
If the `network` structur had access to the bus, wouldn't it be able to implement this internally so we can avoid passing around the function?

I think it'd be simpler given that `getAddrBookPerHeight` doesn't really have any dependencies on `getAddrBookPerHeight` other than bus access.

Reply:
I have refactored things a bit, PTAL and let me know what you think.

The idea is that I am abstracting the `addrBook` sourcing from persistence into its own module so that it's easier to change that specific logic in a centralized place.

In a way this is going to connect with what you mention [here](https://github.com/pokt-network/pocket/pull/374#discussion_r1046585011) 

I am taking "baby steps" to share my WIP but definitely that work should be done separately.

Reply:
👍 Going to resolve this and review the refactored code.